### PR TITLE
Moving the glob parser

### PR DESCRIPTION
**Pull request by mgmoens**
**Problem**:Globbing during shell expansion (see https://github.com/redox-os/ion/issues/258)

**Solution**: I added an extra param to the Normal token which designates it as a normal token that's globable or not.
**Changes introduced by this pull request**:

- Globbing gets moved and probably (isn't) yet parsed 100% correctly. I'm just submitting this PR so I can get in touch with contributors and notify other people of my attempt at trying to fix this.

**Drawbacks**:

Wildcards during quotes don't parse correctly.

**TODOs**: [what is not done yet.]

Make sure wildcards in quotes get parsed as text (if this happened during the previous version).

**Fixes**: [what issues this fixes.]
https://github.com/redox-os/ion/issues/258
**State**: [the state of this PR, e.g. WIP, ready, etc.]
Ready for review
**Blocking/related**: [issues or PRs blocking or being related to this issue.]

**Other**: The author isn't all that used to programming in Rust.

### CODE REVIEW

**Comment by mmstick** [comment]
Thanks for working on this. It's one of the biggest issues with Ion at the moment.

**Comment by mgmoens** [comment]
Is there an IRC channel type thing where you guys hang around? I'd like to ask a couple of questions regarding the code.

**Comment by mmstick** [comment]
I'd be the one to contact regarding Ion, as I wrote all the code here and therefore know all the ins and outs of the implementation. That said, the rest of the Redox team can be found on the team Mattermost chat service: https://chat.redox-os.org/

Or you can ask questions here on this PR or the associated issue, or [send me a direct email](mailto:mmstickman@gmail.com).

Something to look out for with glob expansion is that it will need to return multiple strings, rather than a single string. If glob expressions are found, each expansion should be a unique element so that they work correctly with for loops. The for loop code will also have to be slightly modified, as it's currently using a workaround hack that expands globs specifically for for loops.

**Comment by mgmoens** [comment]
Yup, I thought it was nested globbing that didn't work but it's actually printing out multiple filenames at once, nested globbing works if there's only one dir with one files.

Also, what does join(" ") do? I see that it can be invoked on Small Vecs but it's not in the crates docs.

**Comment by mmstick** [comment]
The `join` method applies to Vec/Array types that contain strings. They just concatenate them with a pattern. In the case of `join(" ")`, that would be joining all strings with a space between each element.

**Comment by mgmoens** [comment]
Found a way to fix the return multiple strings part. 

For the bracket globbing (ie [A-Z]argo.toml => Cargo.toml), can I consider the text inside brackets a glob token if there are no spaces between brackets, no double dots and the string contained is not a single character?

**Comment by mmstick** [comment]
I believe the approach we can use is that if a single-element array does not contain spaces after the [ and before the ], then it is a glob.

**Comment by mgmoens** [comment]
There's still a bug where starting with [] makes the last character spawn it's own string for no reason. The character get's read twice, cus the glob is correct.

I tried using the pipes with examples that did work correctly and it seemed to work fine. Imma add some comments.

Why is Travis failing?

**Comment by mgmoens** [comment]
Ok, I think I'm done. I think it's reviewable at the moment. Fixed the bug I was talking about.

File: src/parser/shell_expand/words.rs [hunk 1]
```diff
@@ -201,14 +201,14 @@ impl<'a> ArrayMethod<'a> {
...
-       
+
         Some("".into()).into_iter().collect()
     }
 }
 
 #[derive(Debug, PartialEq, Clone)]
 pub enum WordToken<'a> {
-    Normal(&'a str),
+    Normal(&'a str,bool),
```
Comment (mmstick) [inline src/parser/shell_expand/words.rs]: Wouldn't it be better to return `Glob(&'a str)` instead?

File: src/parser/shell_expand/words.rs [hunk 2]
```diff
@@ -755,27 +806,71 @@ impl<'a> Iterator for WordIterator<'a> {
...
+                    if rewind == false{
+                        for _ in 0..moves {
+                            iterator.next();
+                        }
+                        self.read+=moves+1;
+                        glob = true;
+                    }
+                    else {
+                        return Some(WordToken::Normal(&self.data[start..self.read],glob));
+                    }
```
Comment (mmstick) [inline src/parser/shell_expand/words.rs]: Should be able to make the above a new function that returns a `Some(&'a str)` so it doesn't need to be copied. Then you can use the map method to convert the returned `Some(&'a str)` into the correct token that you need.

File: src/shell/job.rs [hunk 3]
```diff
@@ -37,10 +37,10 @@ impl Job {
             let mut iterator = self.args.drain();
             expanded.push(iterator.next().unwrap());
             for arg in iterator.flat_map(|argument| expand_string(&argument, expanders, false)) {
-                if arg.contains(|chr| chr == '?' || chr == '*' || chr == '[') {
+                /*if arg.contains(|chr| chr == '?' || chr == '*' || chr == '[') {
```
Comment (mmstick) [inline src/shell/job.rs]: Maybe remove all the commented code now?

Comment (mmstick) [inline src/shell/job.rs]: Perhaps also add a test in the **examples** directory that checks if the following functions correctly? All you do is create the ion file that contains the commands to run, and the corresponding out file that will have the correct output of the standard output of the ion script when execute.

```
echo br*.*
> braces.ion braces.out break.ion break.out
echo [ef]*.ion
> else_if.ion else_if.out fail.ion fail.out fibonacci.ion fibonacci.out fn.ion fn.out for.ion for.out
```

File: src/parser/shell_expand/words.rs (hunk 1 above)
Comment (mgmoens) [inline src/parser/shell_expand/words.rs]: I did it this way to not have an extra if else construct in basically every place that returns a Normal. If you insist I'll gladly change it but I think it's gonna mess up the code.

File: src/shell/job.rs (hunk 3 above)
Comment (mgmoens) [inline src/shell/job.rs]: Added. It's in examples/glob.ion and examples/glob.out.

Comment (mgmoens) [inline src/shell/job.rs]: Done.

File: src/parser/shell_expand/words.rs (hunk 2 above)
Comment (mgmoens) [inline src/parser/shell_expand/words.rs]: I don't really get what you're saying. Basically, in the match (character) part I peek ahead to see if the contents of the string inside the outer [] is globable. If it isn't, I just redirect it to the code that was there before. I know I copy the iterator and that it isn't the most efficient thing, but I need a way to look ahead without break the iterator you use in the next(..) function.

Comment (mmstick) [inline src/parser/shell_expand/words.rs]: What I'm referring to is that you have this logic copied in two different places. One here that ends at line 868, and another copy that ends at line 743.

**Comment by mmstick** [comment]
I'll merge this as is for now.

**Comment by mgmoens** [comment]
I was not available the last couple hours, I'll split up the code regarding 868 etc. tomorrow.

**Comment by mmstick** [comment]
The splitting won't be necessary as I've already done that. Thanks anyway for the PR request! I know this wasn't a particularly easy feature to work on, in addition to getting adjusted to the codebase.

**Pull request merged**